File: common/src/main/java/generations/gg/generations/core/generationscore/world/shop/ShopPresetRegistrySyncPacket.java
package generations.gg.generations.core.generationscore.world.shop;

import generations.gg.generations.core.generationscore.GenerationsCore;
import generations.gg.generations.core.generationscore.network.packets.DataRegistrySyncPacket;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;

import java.util.HashMap;
import java.util.Map;

public class ShopPresetRegistrySyncPacket extends DataRegistrySyncPacket<ShopPreset, ShopPresetRegistrySyncPacket> {
    public ShopPresetRegistrySyncPacket(Map<ResourceLocation, ShopPreset> graphs) {
        super(graphs);
    }

    @Override
    public ResourceLocation getId() {
        return ID;
    }

    @Override
    public void encodeEntry(FriendlyByteBuf buffer, ShopPreset entry) {
        entry.encode(buffer);
    }

    @Override
    public ShopPreset decodeEntry(FriendlyByteBuf buffer) {
        return new ShopPreset(buffer);
    }

    @Override
    public void synchronizeDecoded(Map<ResourceLocation, ShopPreset> entries) {
        ShopPresets.instance().receiveSyncPacket(entries);
    }

    public static final ResourceLocation ID = GenerationsCore.id("shop_registry_sync");

    public static ShopPresetRegistrySyncPacket decode(FriendlyByteBuf buffer) {
        var graph = new ShopPresetRegistrySyncPacket(new HashMap<>());
        graph.decodeBuffer(buffer);
        return graph;
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/shop/ShopPresetRegistrySyncPacket.kt
package generations.gg.generations.core.generationscore.common.world.shop

import generations.gg.generations.core.generationscore.common.GenerationsCore
import generations.gg.generations.core.generationscore.common.network.packets.DataRegistrySyncPacket
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.resources.ResourceLocation

class ShopPresetRegistrySyncPacket(presets: Collection<ShopPreset>) : DataRegistrySyncPacket<ShopPreset, ShopPresetRegistrySyncPacket>(presets) {
    override fun encodeEntry(buffer: RegistryFriendlyByteBuf, entry: ShopPreset) {
        entry.encode(buffer)
    }

    override fun decodeEntry(buffer: RegistryFriendlyByteBuf): ShopPreset {
        return ShopPreset.decode(buffer)
    }

    override fun synchronizeDecoded(entries: Collection<ShopPreset>) {
        ShopPresets.reload(entries.associateBy { it.id })
    }

    override val id: ResourceLocation = ID

    companion object {
        val ID: ResourceLocation = GenerationsCore.id("shop_registry_sync")

        fun decode(buffer: RegistryFriendlyByteBuf): ShopPresetRegistrySyncPacket = ShopPresetRegistrySyncPacket(emptyList()).apply { this.decodeBuffer(buffer) }
    }
}